### PR TITLE
[WIP] Add ROUGE Metric

### LINKED ISSUE #67: Add ROUGE metrics
Splitting this issue out from https://github.com/keras-team/keras-nlp/issues/38.

We should add a ROUGE metric as `keras_nlp.metrics.RougeL`, etc.

### PR DESCRIPTION
**Pull request by abheesht17**
Resolves #67 

- [x] ROUGE-N
- [ ] ROUGE-L

**Important**: @mattdangerw, @chenmoneygithub, the ROUGE-N implementation is very inefficient. I could not find a data structure in TensorFlow which can store tensors (ngrams) as keys (or hash them in a set). So, I have to use `TensorArray` and push an ngram if it does not already belong to the `TensorArray`. Hence,  the push operation is O(n), as opposed to O(1) if we have a set/dictionary data structure in TensorFlow. Please let me know if there is a more efficient way. Currently, ROUGE-N computation makes training very slow. Thank you!

Note: [MutableHashTable](https://www.tensorflow.org/api_docs/python/tf/lookup/experimental/MutableHashTable) does not accept tensors as keys.

Edit: Even ROUGE-L is slow. Is it because I'm looping over the batches during metric computation?

ROUGE-N Notebook: https://colab.research.google.com/drive/1wiNN5KZaClr7qufAu3vzZtx6YECqsTrX?usp=sharing
ROUGE-L Notebook: https://colab.research.google.com/drive/1xchVi4DsG_2gfi8FmO-g9GZS6-7CIFnz?usp=sharing

### CODE REVIEW

**Comment by LukeWood** [comment]
> **Important**: @mattdangerw, @chenmoneygithub, the ROUGE-N implementation is very inefficient. I could not find a data structure in TensorFlow which can store tensors (ngrams) as keys (or hash them in a set). So, I have to use `TensorArray` and push an ngram if it does not already belong to the `TensorArray`. Hence, the push operation is O(n), as opposed to O(1) if we have a set/dictionary data structure in TensorFlow. Please let me know if there is a more efficient way. Currently, ROUGE-N computation makes training very slow. Thank you!

You can actually use a Tensor to index a TensorArray.  so `tensor_array.gather([indices])` works - so we could actually implement a hashing algorithm.  I'm not saying this is worth the effort, but a custom hash table is theoretically possible.

**Comment by abheesht17** [comment]
> > **Important**: @mattdangerw, @chenmoneygithub, the ROUGE-N implementation is very inefficient. I could not find a data structure in TensorFlow which can store tensors (ngrams) as keys (or hash them in a set). So, I have to use `TensorArray` and push an ngram if it does not already belong to the `TensorArray`. Hence, the push operation is O(n), as opposed to O(1) if we have a set/dictionary data structure in TensorFlow. Please let me know if there is a more efficient way. Currently, ROUGE-N computation makes training very slow. Thank you!
> 
> You can actually use a Tensor to index a TensorArray. so `tensor_array.gather([indices])` works - so we could actually implement a hashing algorithm. I'm not saying this is worth the effort, but a custom hash table is theoretically possible.

Nice. Thank you, @LukeWood!

**Comment by chenmoneygithub** [comment]
@abheesht17 Please ping us when this PR is ready for review, thanks!

**Comment by abheesht17** [comment]
Hey, @chenmoneygithub! I'm waiting for @mattdangerw's comments on the two notebooks. We want to optimise the solution. Could you please have a look as well?

ROUGE-N Notebook: https://colab.research.google.com/drive/1wiNN5KZaClr7qufAu3vzZtx6YECqsTrX?usp=sharing
ROUGE-L Notebook: https://colab.research.google.com/drive/1xchVi4DsG_2gfi8FmO-g9GZS6-7CIFnz?usp=sharing

**Comment by abheesht17** [comment]
Closing this PR, and opening two PRs, one each for Rouge-L [https://github.com/keras-team/keras-nlp/pull/122] and ROUGE-N.
